File: junsuk/18주차/Backspace String compare.js
/**
 * @param {string} s
 * @param {string} t
 * @return {boolean}
 */
var backspaceCompare = function(s, t) {
    function process(str){
        const stack = [];
        for (let char of str) {
            if (char === '#') {
                stack.pop();
            } else {
                stack.push(char);
            }
        }
        return stack.join('');
    };

    return process(s) === process(t);
};

//두 문자열 s와 t가 주어졌을 때, 이 두 문자열을 빈 텍스트 편집기에 입력했을 때 내용이 같으면 true를 반환하십시오. '#'은 백스페이스 문자를 의미합니다.
//빈 텍스트를 백스페이스로 지운 후에도 텍스트는 계속 비어 있는 상태로 유지된다는 점에 유의하십시오.

File: junsang/18주차/[LTC] 844_Backspace String Compare/Backspace String Compare.java
class Solution {
    public boolean backspaceCompare(String s, String t) {
        Stack<Character> stack = new Stack<>();

        for(char c : s.toCharArray()){
            if(c != '#'){
                stack.push(c);
            } else if(!stack.isEmpty()){
                stack.pop();
            }
        }

        Stack<Character> stack2 = new Stack<>();

        for(char c : t.toCharArray()){
            if(c != '#'){
                stack2.push(c);
            } else if(!stack2.isEmpty()){
                stack2.pop();
            }
        }

        if(stack.equals(stack2)){
            return true;
        } else {
            return false;
        }
    }
}